service/share/share.go

```diff
@@ -145,16 +146,40 @@ func (s *service) GetSharesByNamespace(ctx context.Context, root *Root, nID name
 	}
 
 	namespacedShares := make([]Share, 0)
+	nodeCh := make(chan []format.Node)
+	errCh := make(chan error)
+	wg := sync.WaitGroup{}
+
 	for _, rootCID := range rowRootCIDs {
-		nodes, err := ipld.GetLeavesByNamespace(ctx, s.dag, rootCID, nID)
-		if err != nil {
-			return nil, err
-		}
+		wg.Add(1)
+		go func(rootCID cid.Cid) {
+			nodes, err := ipld.GetLeavesByNamespace(ctx, s.dag, rootCID, nID)
+			if err != nil {
+				errCh <- err
+			}
+			nodeCh <- nodes
+		}(rootCID)
+	}
 
-		for _, node := range nodes {
-			namespacedShares = append(namespacedShares, node.RawData()[1:])
+	go func() {
+		wg.Wait()
+		close(nodeCh)
+		close(errCh)
+	}()
+
+	go func() {
+		for nodes := range nodeCh {
+			for _, node := range nodes {
+				namespacedShares = append(namespacedShares, node.RawData()[1:])
+			}
+			wg.Done()
 		}
+	}()
+
+	for err := range errCh {
+		return nil, err
 	}
+
```

Comment:
By quick look, there is one bug in here and maybe more: if `ipld.GetLeavesByNamespace` returns an error, `wg.Wait` will never unblock(line 165) and both `errCh` and `nodeCh` won't be closed.

Reply:
Yeah, unfortunately It has one more issue that I haven`t noticed. 
Regarding line 165 I believed that it has to be cleared by GC but it`s not. As a fallback solution I`d like to propose to call `wg.Done` also at line 180 and not return the err but storing it to local variable and return the error. So channels will be closed in this case. The minus is that we have to perform all requests to `ipld.GetLeavesByNamespace` even if error occurred.